### PR TITLE
RIA-3175: Other Needs Section

### PR DESCRIPTION
**Pull request by alexbaeza**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RIA-3175

### Change description ###

Added Other Needs Section which includes:
- Multimedia evidence pages, controllers and content
- Single sex appointment pages, controllers and content
- Private Appointment pages, unit tests and content
- Health conditions pages, unit tests and content
- Past experiences pages, unit tests and content

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: app/controllers/cma-requirements/other-needs/anything-else-question.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    };
+
+    const onSuccess = (answer: boolean) => {
+      if (answer) {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          anythingElse: true
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/anything-else-question.ts:46]: Can we remove this commented out line

File: app/controllers/cma-requirements/other-needs/anything-else-question.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+        return res.redirect(paths.awaitingCmaRequirements.otherNeedsAnythingElseReasons);
+      } else {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          anythingElse: false
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/anything-else-question.ts:55]: and this one.

File: app/controllers/cma-requirements/other-needs/anything-else-reason.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+          error: validationErrors
+        });
+      }
+
+      req.session.appeal.cmaRequirements.otherNeeds = {
+        ...req.session.appeal.cmaRequirements.otherNeeds,
+        anythingElseReason: req.body['reason']
+      };
+
+      // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/anything-else-reason.ts:61]: Should we not be updating CCD here?

File: app/controllers/cma-requirements/other-needs/anything-else-reason.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+          question: {
+            name: 'reason',
+            title: i18n.pages.cmaRequirements.otherNeedsSection.anythingElseReasons.heading,
+            value: savedReason ? savedReason : ''
+          },
+          supportingEvidence: false,
+          timeExtensionAllowed: false,
+          errorList: Object.values(validationErrors),
+          error: validationErrors
+        });
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/anything-else-reason.ts:53]: Not a biggie but could we pull this out into a function we use here and from the get request. Then if we have to add something in the future we don't forget to add it to the model when validation has failed. Was looking at a bug the other day where that had happened and we don't always check the page when we have failed validation so thought that might be a good pattern to use. You'll then have to add the error and errorList. as part of this method I guess.

File: app/controllers/cma-requirements/other-needs/bring-equipment-question.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    };
+
+    const onSuccess = (answer: boolean) => {
+      if (answer) {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          bringOwnMultimediaEquipment: true
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/bring-equipment-question.ts:47]: Remove comment?

File: app/controllers/cma-requirements/other-needs/bring-equipment-question.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+        return res.redirect(paths.awaitingCmaRequirements.otherNeedsSingleSexAppointment);
+      } else {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          bringOwnMultimediaEquipment: false
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/bring-equipment-question.ts:56]: Remove comment?

File: app/controllers/cma-requirements/other-needs/bring-equipment-reason.ts [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+          pageTitle: i18n.pages.cmaRequirements.otherNeedsSection.bringEquipmentReason.title,
+          question: {
+            name: 'reason',
+            title: i18n.pages.cmaRequirements.otherNeedsSection.bringEquipmentReason.heading,
+            value: savedReason ? savedReason : ''
+          },
+          supportingEvidence: false,
+          timeExtensionAllowed: false,
+          errorList: Object.values(validationErrors),
+          error: validationErrors
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/bring-equipment-reason.ts:52]: Same as comment above about pulling the model to a single function

File: app/controllers/cma-requirements/other-needs/bring-equipment-reason.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        });
+      }
+
+      req.session.appeal.cmaRequirements.otherNeeds = {
+        ...req.session.appeal.cmaRequirements.otherNeeds,
+        bringOwnMultimediaEquipmentReason: req.body['reason']
+      };
+
+      // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/bring-equipment-reason.ts:61]: Do we not want to save this?

File: app/controllers/cma-requirements/other-needs/health-conditions-question.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    };
+
+    const onSuccess = (answer: boolean) => {
+      if (answer) {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          healthConditions: true
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/health-conditions-question.ts:46]: Remove comment?

File: app/controllers/cma-requirements/other-needs/health-conditions-question.ts [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+        return res.redirect(paths.awaitingCmaRequirements.otherNeedsHealthConditionsReason);
+      } else {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          healthConditions: false
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/health-conditions-question.ts:55]: Remove comment?

File: app/controllers/cma-requirements/other-needs/health-conditions-reason.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+          pageTitle: i18n.pages.cmaRequirements.otherNeedsSection.healthConditionsReason.title,
+          question: {
+            name: 'reason',
+            title: i18n.pages.cmaRequirements.otherNeedsSection.healthConditionsReason.heading,
+            value: savedReason ? savedReason : ''
+          },
+          supportingEvidence: false,
+          timeExtensionAllowed: false,
+          errorList: Object.values(validationErrors),
+          error: validationErrors
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/health-conditions-reason.ts:52]: Same as comment above about pulling the model to a single function

File: app/controllers/cma-requirements/other-needs/health-conditions-reason.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        });
+      }
+
+      req.session.appeal.cmaRequirements.otherNeeds = {
+        ...req.session.appeal.cmaRequirements.otherNeeds,
+        healthConditionsReason: req.body['reason']
+      };
+
+      // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/health-conditions-reason.ts:61]: Do we not want to save this?

File: app/controllers/cma-requirements/other-needs/multimedia-evidence-question.ts [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    };
+
+    const onSuccess = (answer: boolean) => {
+      if (answer) {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          multimediaEvidence: true
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/multimedia-evidence-question.ts:48]: Remove comment?

File: app/controllers/cma-requirements/other-needs/multimedia-evidence-question.ts [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+        return res.redirect(paths.awaitingCmaRequirements.otherNeedsMultimediaEquipmentQuestion);
+      } else {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          multimediaEvidence: false
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/multimedia-evidence-question.ts:57]: Remove comment?

File: app/controllers/cma-requirements/other-needs/past-experiences-question.ts [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    };
+
+    const onSuccess = (answer: boolean) => {
+      if (answer) {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          pastExperiences: true
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/past-experiences-question.ts:47]: Remove comment?

File: app/controllers/cma-requirements/other-needs/past-experiences-question.ts [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+
+        return res.redirect(paths.awaitingCmaRequirements.otherNeedsPastExperiencesReasons);
+      } else {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          pastExperiences: false
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/past-experiences-question.ts:56]: Remove comment?

File: app/controllers/cma-requirements/other-needs/past-experiences-reason.ts [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+          pageTitle: i18n.pages.cmaRequirements.otherNeedsSection.pastExperiencesReasons.title,
+          question: {
+            name: 'reason',
+            title: i18n.pages.cmaRequirements.otherNeedsSection.pastExperiencesReasons.heading,
+            value: savedReason ? savedReason : ''
+          },
+          supportingEvidence: false,
+          timeExtensionAllowed: false,
+          errorList: Object.values(validationErrors),
+          error: validationErrors
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/past-experiences-reason.ts:52]: Same as comment above about pulling the model to a single function

File: app/controllers/cma-requirements/other-needs/past-experiences-reason.ts [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+        });
+      }
+
+      req.session.appeal.cmaRequirements.otherNeeds = {
+        ...req.session.appeal.cmaRequirements.otherNeeds,
+        pastExperiencesReason: req.body['reason']
+      };
+
+      // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/past-experiences-reason.ts:61]: Do we not need to save this?

File: app/controllers/cma-requirements/other-needs/private-appointment-question.ts [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    };
+
+    const onSuccess = (answer: boolean) => {
+      if (answer) {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          privateAppointment: true
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/private-appointment-question.ts:47]: Remove comment?

File: app/controllers/cma-requirements/other-needs/private-appointment-question.ts [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+
+        return res.redirect(paths.awaitingCmaRequirements.otherNeedsPrivateAppointmentReason);
+      } else {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          privateAppointment: false
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/private-appointment-question.ts:56]: Remove comment?

File: app/controllers/cma-requirements/other-needs/private-appointment-reason.ts [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+          pageTitle: i18n.pages.cmaRequirements.otherNeedsSection.privateAppointmentReason.title,
+          question: {
+            name: 'reason',
+            title: i18n.pages.cmaRequirements.otherNeedsSection.privateAppointmentReason.heading,
+            value: savedReason ? savedReason : ''
+          },
+          supportingEvidence: false,
+          timeExtensionAllowed: false,
+          errorList: Object.values(validationErrors),
+          error: validationErrors
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/private-appointment-reason.ts:52]: Same as comment above about pulling the model to a single function

File: app/controllers/cma-requirements/other-needs/private-appointment-reason.ts [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+        });
+      }
+
+      req.session.appeal.cmaRequirements.otherNeeds = {
+        ...req.session.appeal.cmaRequirements.otherNeeds,
+        singleSexAppointmentReason: req.body['reason']
+      };
+
+      // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/private-appointment-reason.ts:61]: Do we not need to save this?

File: app/controllers/cma-requirements/other-needs/single-sex-appointment-all-female-reason.ts [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+          pageTitle: i18n.pages.cmaRequirements.otherNeedsSection.singleSexAppointmentAllFemale.title,
+          question: {
+            name: 'reason',
+            title: i18n.pages.cmaRequirements.otherNeedsSection.singleSexAppointmentAllFemale.heading,
+            value: savedReason ? savedReason : ''
+          },
+          supportingEvidence: false,
+          timeExtensionAllowed: false,
+          errorList: Object.values(validationErrors),
+          error: validationErrors
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/single-sex-appointment-all-female-reason.ts:52]: Same as comment above about pulling the model to a single function

File: app/controllers/cma-requirements/other-needs/single-sex-appointment-all-female-reason.ts [hunk 24]
```diff
@@ -0,0 +1,90 @@
...
+        });
+      }
+
+      req.session.appeal.cmaRequirements.otherNeeds = {
+        ...req.session.appeal.cmaRequirements.otherNeeds,
+        singleSexAppointmentReason: req.body['reason']
+      };
+
+      // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/single-sex-appointment-all-female-reason.ts:61]: Do we not need to save this?

File: app/controllers/cma-requirements/other-needs/single-sex-appointment-all-male-reason.ts [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+          pageTitle: i18n.pages.cmaRequirements.otherNeedsSection.singleSexAppointmentAllMale.title,
+          question: {
+            name: 'reason',
+            title: i18n.pages.cmaRequirements.otherNeedsSection.singleSexAppointmentAllMale.heading,
+            value: savedReason ? savedReason : ''
+          },
+          supportingEvidence: false,
+          timeExtensionAllowed: false,
+          errorList: Object.values(validationErrors),
+          error: validationErrors
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/single-sex-appointment-all-male-reason.ts:52]: Same as comment above about pulling the model to a single function

File: app/controllers/cma-requirements/other-needs/single-sex-appointment-all-male-reason.ts [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+        });
+      }
+
+      req.session.appeal.cmaRequirements.otherNeeds = {
+        ...req.session.appeal.cmaRequirements.otherNeeds,
+        singleSexAppointmentReason: req.body['reason']
+      };
+
+      // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/single-sex-appointment-all-male-reason.ts:61]: Do we not need to save this?

File: app/controllers/cma-requirements/other-needs/single-sex-appointment-question.ts [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+    };
+
+    const onSuccess = (answer: boolean) => {
+      if (answer) {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          singleSexAppointment: true
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/single-sex-appointment-question.ts:46]: Remove comment?

File: app/controllers/cma-requirements/other-needs/single-sex-appointment-question.ts [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+
+        return res.redirect(paths.awaitingCmaRequirements.otherNeedsSingleSexTypeAppointment);
+      } else {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          singleSexAppointment: false
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/single-sex-appointment-question.ts:55]: Remove comment?

File: app/controllers/cma-requirements/other-needs/single-sex-type-appointment-question.ts [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+    };
+
+    const onSuccess = (answer: boolean) => {
+      if (answer) {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          singleSexTypeAppointment: SexType.ALL_MALE
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/single-sex-type-appointment-question.ts:51]: Remove comment?

File: app/controllers/cma-requirements/other-needs/single-sex-type-appointment-question.ts [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+
+        return res.redirect(paths.awaitingCmaRequirements.otherNeedsAllMaleAppointment);
+      } else {
+        req.session.appeal.cmaRequirements.otherNeeds = {
+          ...req.session.appeal.cmaRequirements.otherNeeds,
+          singleSexTypeAppointment: SexType.ALL_FEMALE
+        };
+
+        // await updateAppealService.submitEvent(Events.EDIT_CMA_REQUIREMENTS, req);
```
Comment (chrisgrimble) [inline app/controllers/cma-requirements/other-needs/single-sex-type-appointment-question.ts:60]: Remove comment?

**Comment by lewis150193** [comment]
Closed

**Pull request closed without merging**